test: tests/test_board_04_routing_regression.py fails on python backend (3/9 nets, was 9/9)

## Context

The pre-existing slow test \`tests/test_board_04_routing_regression.py::TestBoard04OscOutRouting::test_routes_all_nets_on_2l\` is failing with the python backend, asserting:

\`\`\`
E       AssertionError: Board 04 routed only 3/9 nets (expected 9/9).
E       assert 3 >= 9
\`\`\`

The test runs \`kct route ... --backend python --layers 2 --manufacturer jlcpcb --seed 42\` and expects 9/9 nets routed (per issue #2745's OSC_OUT recovery fix).  Today the python backend produces only 3/9 nets on the same recipe.

The C++ backend (now the default) still produces 8/9 on this board (NRST regression filed separately) -- so this is a python-backend-specific regression compounding with the NRST issue.

## Discovery context

Surfaced during board 04 manufacturability verification (#3261, parent #2394). The test is marked \`@slow\` so it does not run on PR-time CI; only nightly slow-tests workflow picks it up. The slow-tests workflow may have been failing silently for some time.

## Acceptance criteria

- [ ] Bisect the python backend to find the change that dropped 9/9 -> 3/9 on this recipe.
- [ ] Either fix the python backend OR update the test to either:
  - Switch the test to \`--backend cpp\` (which is now the documented default), OR
  - Lower the threshold to match current python backend reality and document the gap.
- [ ] Investigate whether the nightly slow-tests workflow has been red on this test for an extended period.

## References
- Parent verification: #3261
- Test originated: PR #2751 (closes #2745)
- Manufacturability cluster: #2834, #2394

--- a/src/kicad_tools/router/grid.py
+++ b/src/kicad_tools/router/grid.py
@@ -294,11 +294,24 @@ def _is_plane_net_pad(pad: "Pad") -> bool:
     validator, allowing trace clips against U2.1 / U2.8 / U2.23 / U2.24
     (Issue #2880).
 
-    The plane-net check is keyed on ``pad.net_name`` (exact match
-    against ``_PLANE_NET_EXACT`` or starts with a member of
-    ``_PLANE_NET_PREFIXES``) so the semantics are consistent
-    regardless of whether the schematic uses ``skip_nets``
-    (``pad.net == 0``) or not (``pad.net != 0``).
+    Issue #3281: The original PR #2931 fix treated ALL ``pad.net == 0``
+    pads as plane pads on the assumption that ``net == 0`` was always
+    the ``skip_nets`` convention.  This silently misclassified
+    no-connect (NC) pins -- which have ``net == 0`` AND ``net_name == ""``
+    inherently, NOT because they were rewritten by ``skip_nets`` -- as
+    plane pads.  On board 04's STM32 LQFP-48 east edge, NC pin 33 sits
+    between SWDIO (pad 34) and an escape channel; classifying it as
+    plane caused the validator to reject the SWDIO escape, dropping
+    board 04 from 9/9 to 4/9 nets routed on the stripped 2L recipe.
+
+    The plane-net check is now keyed exclusively on ``pad.net_name``
+    (exact match against ``_PLANE_NET_EXACT`` or starts with a member
+    of ``_PLANE_NET_PREFIXES``).  The ``net_name`` field is preserved
+    on the pad whether the net was rewritten to ``0`` by ``skip_nets``
+    (``net = 0, net_name = "GND"``) or kept as a real net (``net = 2,
+    net_name = "+3.3V"``), so the same predicate works in both
+    conventions.  NC pads (``net = 0, net_name = ""``) correctly return
+    ``False`` and remain subject to the same-component carve-out.
 
     Args:
         pad: The pad to classify.
@@ -306,11 +319,14 @@ def _is_plane_net_pad(pad: "Pad") -> bool:
     Returns:
         True if the pad's net name matches the plane-net classification.
     """
-    if pad.net == 0:
-        # The skipped-pour-net convention already marks this as a plane.
-        return True
     name = pad.net_name.upper() if pad.net_name else ""
     if not name:
+        # No net_name means either:
+        # (a) NC pin (pad.net == 0 inherently, no schematic net) -- NOT a plane.
+        # (b) A real net but the netlist parser dropped the name -- defensive
+        #     fallback to treating it as signal (non-plane) so we do not
+        #     over-block escape routing.  Pre-#3281 code returned True
+        #     unconditionally for ``pad.net == 0``, which broke board 04.
         return False
     if name in _PLANE_NET_EXACT:
         return True

--- a/tests/router/test_validate_route_clearance_rect.py
+++ b/tests/router/test_validate_route_clearance_rect.py
@@ -287,15 +287,40 @@ class TestIsPlaneNetPad:
         )
         assert _is_plane_net_pad(pad) is False
 
-    def test_skipped_pour_net_is_plane(self) -> None:
-        """``pad.net == 0`` is the skipped-pour-net convention (set by
-        ``io.py`` when the user passes ``--skip-nets GND,+3.3V``).
-        It always classifies as a plane regardless of net_name."""
+    def test_skipped_pour_net_with_name_is_plane(self) -> None:
+        """``pad.net == 0`` with a power/ground ``net_name`` (the skipped-
+        pour-net convention used by ``io.py`` when ``--skip-nets GND,+3.3V``
+        is passed) classifies as a plane.  ``io.py`` preserves the
+        ``net_name`` field even when it rewrites ``net_num`` to ``0``,
+        so the name-based check still fires.
+
+        Issue #3281: Pre-fix code returned ``True`` for ALL ``net == 0``
+        pads regardless of ``net_name``, silently misclassifying NC pins
+        (which have ``net == 0`` AND ``net_name == ""`` inherently) as
+        plane pads.  On board 04's STM32 LQFP-48 this rejected SWDIO /
+        NRST / SWO / BOOT0 escape routes that pass next to NC pins, and
+        dropped 2L routing from 9/9 to 4/9.  The fix narrows the
+        classifier to require an explicit plane-net ``net_name``.
+        """
         pad = Pad(
             x=0.0, y=0.0, width=1.0, height=1.0,
-            net=0, net_name="", ref="U1", pin="1", layer=Layer.F_CU,
+            net=0, net_name="GND", ref="U1", pin="1", layer=Layer.F_CU,
         )
         assert _is_plane_net_pad(pad) is True
+
+    def test_no_connect_pin_is_not_plane(self) -> None:
+        """Issue #3281: NC pins (no schematic connection) inherently have
+        ``net == 0`` and ``net_name == ""``.  They must NOT classify as
+        plane pads -- the same-component carve-out applies to them so
+        signal escapes can route through the NC-channel between active
+        pins (e.g. SWDIO escape east past NC pin 33 on board 04's STM32
+        LQFP-48).
+        """
+        pad = Pad(
+            x=0.0, y=0.0, width=1.0, height=1.0,
+            net=0, net_name="", ref="U1", pin="33", layer=Layer.F_CU,
+        )
+        assert _is_plane_net_pad(pad) is False
 
 
 # -----------------------------------------------------------------------------

--- a/tests/test_board_04_routing_regression.py
+++ b/tests/test_board_04_routing_regression.py
@@ -14,7 +14,8 @@ Per-net ``stall_budget = 3`` prevents thrash.
 
 This test pins the post-fix behavior:
 
-- Board 04 must route 9/9 nets on a default ``kct route`` invocation.
+- Board 04 must route at least 8/9 nets on the stripped 2L recipe
+  (see ``REQUIRED_NETS_ROUTED`` for the current floor).
 - The 2L attempt must succeed (no layer escalation should be needed).
 
 Marked ``@pytest.mark.slow`` (single 2L attempt is ~60-90s; we set a
@@ -28,10 +29,19 @@ the stripped recipe.  Investigation showed the C++ backend on the same
 stripped recipe also produces 4/9, so this is not python-specific; it
 is a broader regression that surfaces here because the test deliberately
 omits ``--micro-via-in-pad-fallback`` and the other production-pipeline
-flags to isolate the #2745 recovery gate.  The router regression is
-tracked in #3281; until that is resolved this test is skipped on the
-python backend so the slow-tests workflow stays green rather than
-red-on-known-gap.
+flags to isolate the #2745 recovery gate.
+
+Issue #3281 (2026-06-07) — bisect identified PR #2931's
+``_is_plane_net_pad`` classifier as the regression source: the
+``pad.net == 0`` shortcut misclassified NC pins (which have
+``net == 0 AND net_name == ""`` inherently, NOT because they were
+rewritten by ``--skip-nets``) as plane pads, blocking same-component
+signal escapes that ran past NC pins on board 04's STM32 LQFP-48 east
+edge.  Narrowing the classifier to require an explicit plane-net
+``net_name`` restored 8/9 on the C++ backend (matching the production-
+recipe baseline).  This test now runs against the C++ backend (the
+documented default) with an 8/9 floor; the residual NRST gap is
+tracked independently per #3281's acceptance criteria.
 """
 
 from __future__ import annotations
@@ -50,12 +60,23 @@ BOARD_DIR = REPO_ROOT / "boards" / "04-stm32-devboard"
 UNROUTED_PCB = BOARD_DIR / "output" / "stm32_devboard.kicad_pcb"
 
 
-# Issue #2745 acceptance criterion: 9/9 nets routed.  Board 04 has 9
-# routable nets after schematic / PCB sync; the OSC_OUT failure was the
-# single net keeping the board at 8/9.  We require 9/9 here so the
-# regression is caught immediately if the BLOCKED_BY_COMPONENT recovery
-# is re-gated.
-REQUIRED_NETS_ROUTED = 9
+# Issue #2745 acceptance criterion: originally 9/9 nets routed.
+#
+# Issue #3281 update: the post-#3128 unrouted PCB (regenerated 2026-05-26
+# to incorporate the micro-via in-pad rescue support) shifted the
+# routing surface enough that NRST no longer routes on the stripped 2L
+# recipe with either backend.  The OSC_OUT regression that motivated
+# #2745 is once again clear after fixing the NC-pin plane-net
+# misclassification at ``router/grid.py::_is_plane_net_pad``
+# (issue #3281), so this test now pins the post-fix floor of 8/9.
+# The residual NRST gap on the stripped 2L recipe is the SAME defect
+# as the C++ backend's NRST gap on the full production recipe
+# (``--mfr jlcpcb-tier1 --auto-fix --auto-layers --auto-mfr-tier
+# --placement-feedback --micro-via-in-pad-fallback``), and is tracked
+# independently per the #3281 acceptance criteria.
+#
+# Board 04 has 9 routable nets after schematic / PCB sync.
+REQUIRED_NETS_ROUTED = 8
 REQUIRED_NETS_TOTAL = 9
 
 
@@ -87,24 +108,31 @@ def _parse_routed_net_count(stdout: str) -> tuple[int, int] | None:
 
 
 @pytest.mark.slow
-@pytest.mark.skip(
-    reason=(
-        "Issue #3268 — python backend regressed from 9/9 to 4/9 on the stripped "
-        "2L recipe (`--no-auto-layers --layers 2 --manufacturer jlcpcb "
-        "--backend python --seed 42`).  The underlying router regression "
-        "(both backends affected) is tracked in #3281; re-enable when that "
-        "is resolved.  The C++-backend production-recipe NRST gap remains "
-        "tracked separately per #3268."
-    )
-)
 class TestBoard04OscOutRouting:
-    """Pin 9/9 routing on board 04 against the #2745 BLOCKED_BY_COMPONENT
-    recovery fix.
+    """Pin >= 8/9 routing on board 04 against the #2745 BLOCKED_BY_COMPONENT
+    recovery fix (and the #3281 NC-pin plane-net misclassification fix).
 
     These tests run the full ``kct route`` CLI as a subprocess to
     exercise the same path the user invokes interactively.  The fixture
     runs once per session; each test asserts a different aspect to keep
     failure attribution sharp.
+
+    Issue #3281: The test was previously skipped because the python
+    backend regressed from 9/9 to 4/9 on this stripped recipe (and the
+    C++ backend matched it at 4/9).  The root cause was the NC-pin
+    misclassification at ``router/grid.py::_is_plane_net_pad`` (post
+    PR #2931): NC pins inherit ``net == 0`` AND ``net_name == ""`` from
+    the netlist parser, but the original classifier returned ``True``
+    for any ``net == 0`` pad, which made the validator reject every
+    same-component-perimeter signal escape that passed an NC pin --
+    blocking SWDIO / SWO / NRST / BOOT0 escapes on board 04's STM32
+    LQFP-48 east edge.  The fix narrows the classifier to require an
+    explicit plane-net ``net_name``.  Both backends now route 8/9 on
+    this recipe (cpp) or 7/9 (python -- pure-Python backend is weaker
+    than C++ and has additional cluster-routing limitations).
+
+    The test now runs on the C++ backend (the documented default per
+    #3268 commentary) so the assertion floor reflects production reality.
     """
 
     @pytest.fixture(scope="class")
@@ -129,7 +157,7 @@ class TestBoard04OscOutRouting:
                 "--timeout",
                 "240",
                 "--backend",
-                "python",
+                "cpp",
             ]
             proc = subprocess.run(
                 cmd,
@@ -154,13 +182,20 @@ class TestBoard04OscOutRouting:
             return proc.stdout
 
     def test_routes_all_nets_on_2l(self, route_stdout: str) -> None:
-        """Board 04 must route all 9 nets on the 2L attempt.
+        """Board 04 must route at least 8/9 nets on the 2L attempt.
 
         Issue #2745: Before the BLOCKED_BY_COMPONENT recovery gate was
         relaxed, OSC_OUT stagnated at 8/9 routed.  After the fix, the
         initial pass's stall recovery sees OSC_OUT (zero placed
         segments) and engages destination-component sibling rip-up
         regardless of the OSC_IN-driven ``overflow = 1``.
+
+        Issue #3281: The same recipe regressed to 4/9 after PR #2931
+        (the validator's NC-pin plane-net misclassification rejected
+        SWDIO / SWO / NRST / BOOT0 escapes).  Narrowing the classifier
+        at ``router/grid.py::_is_plane_net_pad`` restores 8/9 on the
+        C++ backend (matching the production-recipe baseline).  The
+        residual NRST gap is tracked independently.
         """
         parsed = _parse_routed_net_count(route_stdout)
         assert parsed is not None, (

